### ISSUE TITLE
Comparison of linkage results with fastcluster

### ISSUE THREAD

**Issue by nmayorov** [scipy/scipy#6788]
When investigating #6785 I noticed interesting differences between results of `scipy.cluster.hierarchy.linkage` and [fastcluster](http://danifold.net/fastcluster.html).

Here is the script:

```Python
import numpy as np
from scipy.cluster.hierarchy import linkage, _LINKAGE_METHODS
from scipy.spatial.distance import pdist
from fastcluster import linkage as linkage_fc

np.random.seed(0)
X = np.random.randint(0, 2, size=(100, 5))
d = pdist(X, metric='hamming')   # There are few unique distances

max_dist_diff = {}
for method in _LINKAGE_METHODS:
    Z = linkage(d, method=method)
    Z_fc = linkage_fc(d, method=method)
    max_dist_diff[method] = np.max(np.abs(Z[:, 2] - Z_fc[:, 2]))

for key, value in max_dist_diff.items():
    print(key, value)
```
```
single 0.0
weighted 0.1
median 0.0540095268058
complete 0.2
ward 0.302147925263
average 0.0631746031746
centroid 0.026201753277
```

So we see that the distances between clusters being merged don't exactly match. Note that this is the case for both trivial ('median', 'centroid') and nn-chain (all other, except 'single') algorithms. There is no difference for 'single', I suspect because the algorithms are just identical.

The typical picture of the distance differences as the function of the merge number:

![dist_diff](https://cloud.githubusercontent.com/assets/7074023/20306398/cdd9936c-ab5b-11e6-8140-4811b7a479e0.png)

The indices don't match all the way through:

![index](https://cloud.githubusercontent.com/assets/7074023/20306430/fd480f7a-ab5b-11e6-95f5-b0e0d62cecda.png)

------

I think it's not a bug. The explanation that the both functions implement a correct procedure, but because of the initial merge ambiguities (distance ties) at some point the closest distance between clusters begin to vary.

Mainly I base on the fact that, the trivial algorithm is really simple and unlikely to be incorrect, but the distances are different nevertheless. That means that there is no uniquely correct clustering. But a more solid prove/consideration or literature reference would be useful here.

There is a small possibility that something is wrong / different in distance (dissimilarity) computations. I think it's worth to put it here for a little discussion.

@richardtsai, as an author of the core clustering code in Cython, could you please check this report and let us know what do you think?

**Comment by richardtsai** [comment]
Hi Nikolay,

I currently don't have a computer set up with a proper environment around
so I don't have enough details about this. But I do have a few observations
at a first glance.

1. As far as I know not all linkage methods are suitable for non-Euclidean
distance matrices, at least Ward is not.
2. As `linkage` and `nn_chain` share only the distance update functions,
and a update function as simple as the one used in the 'complete' method is
less likely to be incorrectly implemented, I suppose fastcluster uses a
different distance updating/comparison strategy.
3. The results diverse even in the first merge, which does not involve
distance calculation. This seems to be a result of the difference in the
comparison step.

I will do more experiments as soon as I can.

BTW, have you try verifying the results by hand or by plotting the
dendrogram?

Nikolay Mayorov <notifications@github.com>于2016年11月15日 周二21:07写道：

**Comment by nmayorov** [comment]
> 1. As far as I know not all linkage methods are suitable for non-Euclidean
distance matrices, at least Ward is not.

Good catch, I somehow forgot about it. I recomputed with Euclidean distance:

```
complete 0.414213562373
median 0.11803398875
weighted 0.207106781187
centroid 0.105683618151
ward 0.619882321647
single 0.0
average 0.118346732107
```

So conceptually it is the same picture.

> The results diverse even in the first merge, which does not involve
distance calculation. This seems to be a result of the difference in the
comparison step.

I think what you mean is different picks in case of tie dissimilarities, right? It seems totally legitimate to select any 2 clusters with minimum dissimilarity between them.

> BTW, have you try verifying the results by hand or by plotting the
dendrogram?

The situation is that for small number of nodes the merge distances are all the same (and dendrograms are identical), so it's hard to find a small problematic example to go through by hand. 

The resulting dendrograms are visibly different.

-----------

I made another small test. Now I run all clustering in scipy by the basic `_hierarchy.linkage` algorithm. The result:

```
weighted 0.0
single 0.0
complete 0.0
average 0.0232466795209
centroid 0.105683618151
ward 0.429156448034
median 0.11803398875
```

So now some "simpler" schemes match, but other still don't match. I'm starting to get quite confused here.

For now I will try to determine why my nn_chain gives different result say for 'complete' linkage, presumably fastcluster should use the same algorithm.

**Comment by nmayorov** [comment]
So I was able to make "nn_chain" work the same for 'complete' and 'weighted' schemes. I made one potentially rather significant fix, but unrelated to the problem. And another fix was conventional (was it?): when merging clusters, throw out a cluster with the minimum index (actually rather index of a representative). I actually believe it is a conventional fix, when I change it to "throw out a cluster with the maximum index" I again observe differences, but there is no reason to prefer one way or another.

What is the deal with 'ward' and 'average' --- still not sure.

Generally I'm inclined to think, that the difference we see is not a mistake, but just different viable clusterings.

**Comment by nmayorov** [comment]
So I compared thoroughly what happens inside fastcluster and inside nn_chain. Turned out that the reason was mere perturbations in floating point computations, and as distance updates are more complicated in 'ward' and 'average' (divisions and multiplications are involved) --- we see such deviations.

So my conclusion that we don't have any real problems here. I found one potential bad case in nn_chain and will send a pull request soon.
